dreamer/models/agent.py

```diff
@@ -0,0 +1,123 @@
+from dreamer.models.observation import ObservationDecoder, ObservationEncoder
+from dreamer.models.rnns import RSSMState, RSSMRepresentation, RSSMTransition, RSSMRollout, get_feat
+
+
+class AgentModel(nn.Module):
+    def __init__(
+            self,
+            output_size,
+            stochastic_size=30,
+            deterministic_size=200,
+            hidden_size=200,
+            image_shape=(3, 64, 64),
+            action_hidden_size=200,
+            action_layers=3,
+            action_dist='tanh_normal',
+            reward_shape=(1,),
+            reward_layers=3,
+            reward_hidden=200,
+            value_shape=(1,),
+            value_layers=3,
+            value_hidden=200,
+            dtype=torch.float,
+            **kwargs,
+    ):
+        super().__init__()
+        self.observation_encoder = ObservationEncoder(shape=image_shape)
+        encoder_embed_size = self.observation_encoder.embed_size
+        decoder_embed_size = stochastic_size + deterministic_size
+        self.observation_decoder = ObservationDecoder(embed_size=decoder_embed_size, shape=image_shape)
+        self.transition = RSSMTransition(output_size, stochastic_size, deterministic_size, hidden_size)
+        self.representation = RSSMRepresentation(self.transition, encoder_embed_size, output_size, stochastic_size,
+                                                 deterministic_size, hidden_size)
+        self.rollout = RSSMRollout(self.representation, self.transition)
+        feature_size = stochastic_size + deterministic_size
+        self.action_size = output_size
+        self.action_dist = action_dist
+        self.action_decoder = ActionDecoder(output_size, feature_size, action_hidden_size, action_layers, action_dist)
+        self.reward_model = DenseModel(feature_size, reward_shape, reward_layers, reward_hidden)
+        self.value_model = DenseModel(feature_size, value_shape, value_layers, value_hidden)
+        self.dtype = dtype
+        self.stochastic_size = stochastic_size
+        self.deterministic_size = deterministic_size
+
+    def forward(self, observation: torch.Tensor, prev_action: torch.Tensor = None, prev_state: RSSMState = None):
+        state = self.get_state_representation(observation, prev_action, prev_state)
+        action, action_dist = self.policy(state)
+        value = self.value_model(get_feat(state))
+        reward = self.reward_model(get_feat(state))
+        return action, action_dist, value, reward, state
+
+    def policy(self, state: RSSMState):
+        feat = get_feat(state)
+        action_dist = self.action_decoder(feat)
+        if self.action_dist == 'tanh_normal':
+            if self.training:  # use agent.train(bool) or agent.eval()
+                action = action_dist.sample()
+            else:
+                action = action_dist.mode()
+        else:
+            # cannot propagate gradients with one hot distribution
+            action = action_dist.sample()
```

Comment:
I don't understand why all 3 if cases have the same line.  Are the comments notes about things to add in?